platform-sdk/swirlds-common/src/main/java/com/swirlds/common/threading/CountUpLatch.java

```diff
@@ -0,0 +1,187 @@
+     */
+    public void set(final long count) {
+        final long result = currentCount.accumulateAndGet(count, SET_COUNT);
+        if (result != count) {
+            throw new IllegalArgumentException(
+                    "Can't set the count to a lower value. Previous = " + result + ", provided = " + count);
+        }
+        lock.lock();
+        try {
+            condition.signalAll();
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * Increment the count by the given delta.
+     * <p>
+     * Methods that update the count are not mutually thread safe. Concurrent threads should never attempt to update
+     * the count. It is, however, safe for many threads to concurrently call methods that query the count.
+     *
+     * @param delta the amount to increment the count by
+     * @throws IllegalArgumentException if the delta is negative
+     */
+    public void add(final long delta) {
+        if (delta < 0) {
+            throw new IllegalArgumentException("Cannot add a negative delta (" + delta + ")");
+        }
+        currentCount.addAndGet(delta);
+
+        lock.lock();
+        try {
+            condition.signalAll();
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * Wait until the count reaches the given value.
+     *
+     * @param count the count to wait for
+     * @throws InterruptedException if interrupted while waiting
+     */
+    public void await(final long count) throws InterruptedException {
+        if (currentCount.get() >= count) {
+            return;
+        }
+
+        while (true) {
+            lock.lock();
+            try {
+                if (currentCount.get() >= count) {
+                    return;
+                }
+                condition.await();
+            } finally {
+                lock.unlock();
+            }
+        }
+    }
+
+    /**
+     * Wait until the count reaches the given value.
+     *
+     * @param count      the count to wait for
+     * @param timeToWait the maximum time to wait
+     * @return true if the count reached the given value, false if the time to wait expired
+     * @throws InterruptedException if interrupted while waiting
+     */
+    @SuppressWarnings("ResultOfMethodCallIgnored")
+    public boolean await(final long count, final Duration timeToWait) throws InterruptedException {
+        if (currentCount.get() >= count) {
+            return true;
+        }
+
+        final Instant start = Instant.now();
+
+        while (true) {
+            final Instant now = Instant.now();
+            final Duration elapsed = Duration.between(start, now);
+            if (isLessThan(timeToWait, elapsed)) {
+                // we are out of time
+                return currentCount.get() >= count;
+            }
+            final long remainingMillis = timeToWait.minus(elapsed).toMillis();
+
+            lock.lock();
+            try {
+                if (currentCount.get() >= count) {
```

Comment:
with conditions, it's almost always better to "if(condition) then wait" rather than "if(!condition) then return".  Among other things it makes the intent much clearer for the next person to read it, and avoids accidentally waiting unnecessarily.

Reply:
I've changed this so that the condition is in the while loop.